### PR TITLE
feat!: Identity master key can be used only to update identity

### PR DESCRIPTION
**Pull request by shuplenkov**
## Issue being fixed or feature implemented

From security considerations, the master key should be used only for updating and disabling identity. For signing other state transitions we need to create another authorization key with security level - high. 

## What was done?

Changed state transition signature validation

## How Has This Been Tested?

With tests

## Breaking Changes

"High" security level keys must be used insted of "master"

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: packages/js-dpp/lib/stateTransition/validation/validateStateTransitionIdentitySignatureFactory.js [hunk 1]
```diff
@@ -57,6 +59,16 @@ function validateStateTransitionIdentitySignatureFactory(
       return result;
     }
 
+    if (
```
Comment (shumkov) [inline packages/js-dpp/lib/stateTransition/validation/validateStateTransitionIdentitySignatureFactory.js:62]: Shouldn't we use here `getKeySecurityLevelRequirement()` as well?

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/dashmate/configs/schema/configJsonSchema.js [hunk 2]
```diff
@@ -511,8 +515,12 @@ module.exports = {
               type: ['string', 'null'],
               minLength: 1,
             },
+            highPublicKey: {
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/dashmate/configs/schema/configJsonSchema.js:518]: We need master keys to create system identities in Drive. Agreed to rename "high" with "second"

File: packages/js-dpp/lib/stateTransition/validation/validateStateTransitionIdentitySignatureFactory.js [hunk 3]
```diff
@@ -51,18 +65,62 @@ function validateStateTransitionIdentitySignatureFactory(
...
+      }
+    } catch (e) {
+      if (e instanceof StateTransitionIsNotSignedError) {
+        result.addError(
+          new StateTransitionIsNotSignedConsensusError(stateTransition),
+        );
+      } else if (e instanceof PublicKeyMismatchError) {
+        result.addError(
+          new PublicKeyMismatchConsensusError(e.getPublicKey()),
+        );
```
Comment (shumkov) [inline packages/js-dpp/lib/stateTransition/validation/validateStateTransitionIdentitySignatureFactory.js:90]: This is impossible due to check above so we can skip this one.

File: packages/js-dpp/lib/stateTransition/validation/validateStateTransitionIdentitySignatureFactory.js [hunk 4]
```diff
@@ -51,18 +65,62 @@ function validateStateTransitionIdentitySignatureFactory(
...
+      if (!signatureIsValid) {
+        result.addError(
+          new InvalidStateTransitionSignatureConsensusError(stateTransition),
+        );
+      }
+    } catch (e) {
+      if (e instanceof StateTransitionIsNotSignedError) {
+        result.addError(
+          new StateTransitionIsNotSignedConsensusError(stateTransition),
+        );
```
Comment (shumkov) [inline packages/js-dpp/lib/stateTransition/validation/validateStateTransitionIdentitySignatureFactory.js:86]: This is checking in basic validation so we can skip it

File: packages/js-dpp/lib/stateTransition/validation/validateStateTransitionIdentitySignatureFactory.js [hunk 5]
```diff
@@ -51,18 +65,62 @@ function validateStateTransitionIdentitySignatureFactory(
...
+      } else if (e instanceof PublicKeyMismatchError) {
+        result.addError(
+          new PublicKeyMismatchConsensusError(e.getPublicKey()),
+        );
+      } else if (e instanceof InvalidIdentityPublicKeyTypeError) {
+        result.addError(
+          new InvalidIdentityPublicKeyTypeConsensusError(e.getPublicKeyType()),
+        );
```
Comment (shumkov) [inline packages/js-dpp/lib/stateTransition/validation/validateStateTransitionIdentitySignatureFactory.js:94]: This is impossible due to check above

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/js-dpp/lib/errors/consensus/signature/InvalidIdentityPublicKeyTypeError.js [hunk 6]
```diff
@@ -2,24 +2,20 @@ const AbstractSignatureError = require('./AbstractSignatureError');
 
 class InvalidIdentityPublicKeyTypeError extends AbstractSignatureError {
```
Comment (shumkov) [inline packages/js-dpp/lib/errors/consensus/signature/InvalidIdentityPublicKeyTypeError.js:3]: Do we use it anywhere?

File: packages/js-dpp/lib/errors/consensus/signature/PublicKeyIsDisabledError.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
+const AbstractSignatureError = require('./AbstractSignatureError');
+
+class PublicKeyIsDisabledError extends AbstractSignatureError {
+  /**
+   *
+   * @param {IdentityPublicKey} publicKey
```
Comment (shumkov) [inline packages/js-dpp/lib/errors/consensus/signature/PublicKeyIsDisabledError.js:6]: We can't pass objects like that into consensus error due to serialization. It must be types supported by cbor. I would recommend to pass key ID

File: packages/js-dpp/lib/stateTransition/AbstractStateTransitionIdentitySigned.js [hunk 8]
```diff
@@ -121,6 +122,16 @@ class AbstractStateTransitionIdentitySigned extends AbstractStateTransition {
    * and purpose to sign this state transition
    */
   verifyPublicKeyLevelAndPurpose(publicKey) {
+    if (
```
Comment (shumkov) [inline packages/js-dpp/lib/stateTransition/AbstractStateTransitionIdentitySigned.js:125]: Shouldn't it be vice-versa? I mean it should go from what state transition require but not from what key we are passing. Isn't?

File: packages/js-dpp/lib/stateTransition/AbstractStateTransitionIdentitySigned.js [hunk 9]
```diff
@@ -220,18 +231,18 @@ class AbstractStateTransitionIdentitySigned extends AbstractStateTransition {
    * @return {number}
    */
   getKeySecurityLevelRequirement() {
-    return IdentityPublicKey.SECURITY_LEVELS.MASTER;
+    return IdentityPublicKey.SECURITY_LEVELS.HIGH;
```
Comment (shumkov) [inline packages/js-dpp/lib/stateTransition/AbstractStateTransitionIdentitySigned.js:234]: We need to set master to IdentityUpdateTransition class

**Review by shumkov — CHANGES_REQUESTED** [review]
signatureSecurityLevelRequirement in DataContract shouldn't allow to set MASTER

File: packages/js-dpp/lib/errors/consensus/signature/MasterPublicKeySecurityLevelIsNotAllowedError.js [hunk 10]
```diff
@@ -0,0 +1,12 @@
+const AbstractSignatureError = require('./AbstractSignatureError');
```
Comment (shumkov) [inline packages/js-dpp/lib/errors/consensus/signature/MasterPublicKeySecurityLevelIsNotAllowedError.js:1]: Is not using anywhere

File: packages/js-dpp/lib/errors/consensus/signature/InvalidIdentityPublicKeyTypeError.js (hunk 6 above)
Comment (shuplenkov) [reply packages/js-dpp/lib/errors/consensus/signature/InvalidIdentityPublicKeyTypeError.js:3]: Yes

File: packages/platform-test-suite/test/e2e/masternodeRewardShares.spec.js [hunk 11]
```diff
@@ -92,6 +94,68 @@ describe('Masternode Reward Shares', () => {
       await client.platform.identities.topUp(masternodeIdentity.getId(), 7000);
     });
 
+    it('should add high security level public key', async () => {
+      const account = await client.platform.client.getWalletAccount();
```
Comment (shumkov) [inline packages/platform-test-suite/test/e2e/masternodeRewardShares.spec.js:98]: It's not a test so let's move to `before`

**Review by shumkov — APPROVED** [review]
Good job! 👍

**Pull request merged**